replica/table.cc

```diff
@@ -2983,9 +2983,67 @@ bool table::requires_cleanup(const sstables::sstable_set& set) const {
     }));
 }
 
-future<> table::cleanup_tablet(locator::tablet_id) {
-    co_await flush();
-    // FIXME: Remove sstables
+future<> compaction_group::wipe() {
+    class compaction_group_cleaner : public row_cache::external_updater_impl {
+        table& _t;
+        compaction_group& _cg;
+        const lw_shared_ptr<sstables::sstable_set> _empty_main_set;
+        const lw_shared_ptr<sstables::sstable_set> _empty_maintenance_set;
+        std::vector<sstables::shared_sstable>& _all_sstables;
+        std::optional<compaction_manager::compaction_reenabler> _compaction_guard;
+    private:
+        lw_shared_ptr<sstables::sstable_set> empty_sstable_set() const {
+            return make_lw_shared<sstables::sstable_set>(_t._compaction_strategy.make_sstable_set(_t._schema));
+        }
+    public:
+        explicit compaction_group_cleaner(compaction_group& cg, std::vector<sstables::shared_sstable>& all_sstables)
+                : _t(cg._t)
+                , _cg(cg)
+                , _empty_main_set(empty_sstable_set())
+                , _empty_maintenance_set(empty_sstable_set())
+                , _all_sstables(all_sstables)
+        {
+        }
+        virtual future<> prepare() override {
+            _compaction_guard.emplace(co_await _t.get_compaction_manager().stop_and_disable_compaction(_cg.as_table_state()));
+            co_await _cg.flush();
```

Comment:
We must also prevent synchronize with writes so that they don't create and flush new memtables. Current fencing based on topology version only ensures that old writes on this replica will fail, but they may still be running. 

We should either synchronize with them here, or make fencing more strict.

Reply:
> We must also prevent synchronize with writes so that they don't create and flush new memtables. Current fencing based on topology version only ensures that old writes on this replica will fail, but they may still be running.
> 
> We should either synchronize with them here, or make fencing more strict.

we have an async gate which is checked when applying new mutations and submitting new async operations. I was planning to make it per tablet next, but I can do it now.

If we fail on cleanup phase with an exception, do we rollback the migration?

Reply:
> If we fail on cleanup phase with an exception, do we rollback the migration?

No, you can assume there will be no rollback, only retry.

Reply:
Great.

Reply:
@tgrabiec Tomek, I am curious why you specified only writes, but not reads.

Reply:
If in-flight writes are possible

Does it mean we may have missed those writes into the tablet that was migrated away into a new destination?

Or those in-flight writes are also being written into new destination (both old and new)? If I remember correctly, that's it. Please just confirm.

That means I can just ignore those writes during cleanup phase, right?

Reply:
Never mind,

```
4. write_both_read_new

    Precondition: New tablet replicas contain all the writes which reached the matching leaving tablet replicas
    before step 3.
```

Reply:
> @tgrabiec Tomek, I am curious why you specified only writes, but not reads.

Because reads don't have side-effects. Note that we're talking about fenced requests, so coordinators will not get a positive response. 

> Does it mean we may have missed those writes into the tablet that was migrated away into a new destination?

No, the topology change coordination algorithm ensures that ACKed writes will not be lost. Coordinators will write to new replicas.

> That means I can just ignore those writes during cleanup phase, right?

Requests should be failed.

Note that even if we are not migrating, writes may be failed for various reasons and that doesn't mean data loss.

Reply:
> > @tgrabiec Tomek, I am curious why you specified only writes, but not reads.
> 
> Because reads don't have side-effects. Note that we're talking about fenced requests, so coordinators will not get a positive response.
> 

To clarify some more, such writes could cause memtable flushes which leave sstables behind. But we want cleanup to drop all such sstables. This would violate expectations from cleanup, which can lead to data resurrection.

Reply:
It seems we cannot unconditionally fail those in-flight writes though, they might be somewhere between commitlog's apply and table's memtable apply. otherwise we'll leak commitlog entries after cleanup. 

we can fail requests that will still append into commitlog, and we'd have to drain all writes that are past that point, meaning we'll wait for them to apply into memtable, so flush can release commitlog segments.

Reply:
We should only start removal after we switched to a topology that doesn't contain the tablet being removed. At that point in time, we don't have any in-flight writes, and no new ones can be started (@gusev-p will fence them on entry).

Reply:
> so flush can release commitlog segments.

But flush doesn't guarantee that commitlog segments are released, since they may be pinned by other memtables. We need a new solution for the commitlog. See https://github.com/scylladb/scylladb/issues/4734.

Reply:
It's pretty confusing since we can't use mutation timestamps here.

We could have a side data structure that collects valid (replay position range, token range) tuples. When a node gains a range, it adds this tuple (with open right hand side for replay position range). When a node loses a range, it closes the replay position range. When a replay position range drops off the commitlog, the tuple can be removed.

Reply:
We might decide to seal commitlog segments on range movements (losing up to 32MB) and so don't need to deal with full replay positions, just segment numbers.

Reply:
Note that commitlog is not prepared for tablets, as RPs may overlap across memtables (one in each group) in the same table. And on flush completion, commitlog would discard data of multiple groups.  Can we extend segment format to be aware of tablet uuids? Perhaps a pair of table_id, and optional tablet_id.

Reply:
We actually have a per table's refcount in each segment during discarding, so the side effect is that segments are only released once all tablets have their memtables flushed (something we have to fix anyway). but I wonder if RPs overlapping might be a problem elsewhere. According to Callee, CL is a lot dependent on monotonically increasing RPs, and that overlapping is not allowed.

Reply:
If a tablet has a UUID of its own, can the destination tablet be determined before adding the mutation to CL? If so, simply use the tabled UUID instead of table UUID. On replay, we need to be able to map these UUID again of course.